lib.js

```diff
@@ -7,7 +7,71 @@
  * @param {Filter} filter
  */
 function Iterator(friends, filter) {
-    console.info(friends, filter);
+    // console.info(friends, filter);
+
+    if (!(filter instanceof Filter)) {
+        throw new TypeError();
+    }
+
+    this.friends = getFriends(friends).filter((arg) => filter.filter(arg));
+    this.position = 0;
+}
+
+Iterator.prototype.done = function () {
+    return this.position >= this.friends.length;
+};
+
+Iterator.prototype.next = function () {
+    if (this.done()) {
+        return null;
+    }
+
+    return this.friends[this.position++];
+};
+
+function compareFriends(friend1, friend2) {
```

Comment:
Может тогда 
```suggestion
function compareFriends(a, b) {
```
в имени функции и так понятно, что сравнивается.